samples/bluetooth/mesh/src/main.c

```diff
@@ -315,8 +314,8 @@ static int gen_onoff_send(bool val)
 		return -ENOENT;
 	}
 
-	BT_MESH_MODEL_BUF_DEFINE(buf, OP_ONOFF_SET_UNACK, 2);
-	bt_mesh_model_msg_init(&buf, OP_ONOFF_SET_UNACK);
+	BT_MESH_SIG_MODEL_OP_2_BUF_INIT(buf, OP_ONOFF_SET_UNACK, 2);
```

Comment:
@PavelVPV @LingaoM the general goal of this PR is nice. Simplifying message initialisation like this is good. However, this feels like one step forward, but also another step backward. The original idea was that there should be only a single place where you need concerned with the size of the op code, and that's when you define the `OP_*` macros for each message, i.e. that's where you need to know which `BT_MESH_MODEL_OP_*()` macro to use. Everywhere else the code shouldn't need to be concerned with this, rather all size information can be derived from the value itself. Now, it looks like you're adding a second place where the user needs to be conscious of what the op size is, i.e. when initializing messages like this. It's not just bad that you're increasing the amount of places, but you're also inviting potential bugs if the chosen `BT_MESH_SIG_MODEL_OP_*_BUF_INIT()` doesn't match the actual size of the opcode that you give as a second parameter to it.

So I'd suggest to think just a little bit more about this, and if we could avoid the problem described above, while still succeeding to simplify the message initialisation process.

Reply:
> but you're also inviting potential bugs if the chosen BT_MESH_SIG_MODEL_OP_*_BUF_INIT() doesn't match the actual size of the opcode that you give as a second parameter to it.

I don't think so, if you see my implementation, if actual size not match parameter, the build assert will report it.

Reply:
And why add separate macro for opcode, please see: https://github.com/zephyrproject-rtos/zephyr/pull/67662#discussion_r1454383625

Reply:
I am not sure it is not possible (at least I don't know how to do this at the moment) to use a single macro as this implementation relies on `IS_EQ` macro that does some magic which is seems (to me) impossible to do with a comparison (which is what we need for single macro).

Reply:
I also tried to use a single macro to implement it and did many experiments, but in the end it didn't work at all😔.

Reply:
Have you considered changing the way the opcodes are defined? Maybe there's a way to retain the size information in a way that the compiler still thinks is build-time resolvable? Ideally this would still be just one macro, but in the worst case you'd have two macros per opcode: one for the value and another for the size - this way you don't have to hardcode the size as a literal integer when initializing messages.

Reply:
> but in the worst case you'd have two macros per opcode: one for the value and another for the size - this way you don't have to hardcode the size as a literal integer when initializing messages.

This is a big modification, but I think it is not friendly, especially for vendors opcode.

Reply:
> Have you considered changing the way the opcodes are defined? Maybe there's a way to retain the size information in a way that the compiler still thinks is build-time resolvable? 

I have thought about it too, but unfortunately I have no idea.

Reply:
@jhedberg Ping

Reply:
Also ask ChatGPT, but the response not ideal.

Reply:
> Ideally this would still be just one macro, but in the worst case you'd have two macros per opcode: one for the value and another for the size - this way you don't have to hardcode the size as a literal integer when initializing messages.

Done.

Reply:
> > Ideally this would still be just one macro, but in the worst case you'd have two macros per opcode: one for the value and another for the size - this way you don't have to hardcode the size as a literal integer when initializing messages.
> 
> Done.

It looks like you misunderstood what I was trying to say. Creating defines like `BT_MESH_MODEL_OP_LEN_2` doesn't help at all, since this still requires the user to know the length of each opcode. What I meant was, that in the worst case we'd have a way to look up (at build-time) the length of each opcode either automatically (using some macrobatics) or manually through a define for each opcode. In the "manual" case you'd have two defines for each opcode: `OP_FOO` (what the code has had until now) and then `OP_FOO_LEN`. That way you still don't need to care what the length is, since you just append `_LEN` to the opcode macro. That said, I really think there should be a way to automate this: if, when defining opcodes, we generate the length information in a suitable way, it should be possible to programmatically (by the preprocessor) look it up when initialising messages.

Reply:
@jhedberg Thanks for your remind, that cool, done.